--- user.rb	2018-09-11 14:28:22.000000000 +0100
+++ new_user.rb	2018-09-11 14:35:51.000000000 +0100
@@ -57,6 +57,14 @@
           @gl_user = find_user
         end
 
+        def oauth_config
+          Gitlab::Auth::OAuth::Config
+        end
+
+        def admin_groups_enabled?
+           !oauth_config.admin_groups.nil?
+        end
+
         def find_user
           user = find_by_uid_and_provider
 
@@ -65,6 +73,13 @@
 
           user.external = true if external_provider? && user&.new_record?
 
+          if user
+            # Check if there is overlap between the user's groups and the external groups
+            # setting then set user as external or internal.
+            log.info "!!!!!!!!!!!!!!!! GROUP INFO: #{auth_hash.to_json} #{oauth_config.to_json}"
+            user.admin = !(auth_hash.groups & oauth_config.admin_groups).empty? if admin_groups_enabled?
+          end
+
           user
         end
 
