--- user.rb.org	2019-02-22 11:41:15.000000000 +0000
+++ user.rb	2019-02-22 11:42:44.000000000 +0000
@@ -47,12 +47,36 @@
           gl_user.block if block_after_save
 
           log.info "(#{provider}) saving user #{auth_hash.email} from login with admin => #{gl_user.admin}, extern_uid => #{auth_hash.uid}"
+
+          auth_hash.groups.each do |group|
+            log.info "Processing #{group} for user #{auth_hash.email}"
+            gl_group = Group.find_by_path_or_name(group)
+            if gl_group
+              if gl_group.member?(gl_user)
+                log.info "#{auth_hash.email} is already a member of #{group}"
+              else
+                log.info "adding user #{auth_hash.email} to the group #{group}"
+                gl_group.add_user(gl_user, :developer)
+              end
+            else
+              log.info "No matching gitlab group found for #{group}"
+            end
+          end
+
           gl_user
         rescue ActiveRecord::RecordInvalid => e
           log.info "(#{provider}) Error saving user #{auth_hash.uid} (#{auth_hash.email}): #{gl_user.errors.full_messages}"
           [self, e.record.errors]
         end
 
+        def oauth_config
+          Gitlab::Auth::OAuth::Config
+        end
+
+        def admin_groups_enabled?
+           !oauth_config.admin_groups.nil?
+        end
+
         def gl_user
           return @gl_user if defined?(@gl_user)
 
@@ -67,6 +91,12 @@
 
           user.external = true if external_provider? && user&.new_record?
 
+          if user
+            # Check if there is overlap between the user's groups and the external groups
+            # setting then set user as external or internal.
+            user.admin = !(auth_hash.groups & oauth_config.admin_groups).empty? if admin_groups_enabled?
+          end
+
           user
         end
 
